add article about writing Stratospheric (#6)
rieckpil/progmot: content/post/writing-a-book-remotely.md

Both Spring Boot and AWS are more or less mature, at least in the areas that we talk about: SQS, SES, RDS, web development, etc. Still, there's much going on for the library that we're using for integrating AWS services ([Spring Cloud AWS](https://awspring.io/)). Furthermore, we also learn even better best practices or different ways of doing things.

Thanks to the agile nature of Leanpub and given the fact that we didn't have to ship a physical book to our readers, we are still able to update the eBook post-delivery. Leanpub has a feature to notify the readers about a new version that they can then download for free.

While we had no strict roadmap in mind, we all agreed to keep this book up-to-date to avoid a stale book.

Whenever there's a new update for our eBook, we also send out a small release information email to our newsletter to give a short overview of the changes:

![Stratospheric Post Release Mailing List Update](/img/stratospheric-ebook/stratospheric-post-release-mailing-list-update.png#center "Stratospheric Post Release Mailing List Update")

> On the monetary side, we don't get rich from selling this book. It's some additional pocket money for us in the range of $150-300 for every author each month.

Reviewer: ```suggestion
On the financial side, we don't get rich from selling this book. It's some additional pocket money for us in the range of $150-300 for every author each month.
```

> To increase the visibility of our eBook, we listed it on Amazon Kindle. While Amazon's royalty rules are not that generous compared to Leanpub, we primarily list on Amazon to get in front of more developers. Amazon is one of the most significant search interfaces after Google and YouTube.

Reviewer: ```suggestion
To increase the visibility of our eBook, we listed it on Amazon Kindle. While Amazon's royalty rules are not that generous compared to Leanpub, we primarily publish on Amazon as well to get in front of more developers. Amazon is one of the most significant search interfaces after Google and YouTube.
```

> We thought about a printed eBook version, but for now, we postponed this move as we believe our audience is more interested in reading technical books on their devices anyway. However, (at least I) want to have a physical copy of this book on my bookshelf somewhere in the future.

Reviewer: ```suggestion
We thought about a printed eBook version, but for now, we postponed this move as we believe our audience is more interested in reading technical books on their devices anyway. However, at least I'd like to have a physical copy of this book on my bookshelf sometime in the future.
```

Based on the feedback received so far, we also decided to build an online course on top of the eBook to provide some more in-depth coverage of various areas. Given my [existing experience with creating and launching technical online courses](https://progmot.com/post/creating-and-selling-my-first-online-course/), this will be a great opportunity to apply my lessons learned from my first courses.

> Getting to the end of the book, some publishers started to reach out to us. They wanted to publish Stratospheric from their publishing house and bring it to various marketplaces.  The additional outreach may be significant, but the royalties are diminishing compared to Leanpub, especially if we have to split them into three equal parts.

Reviewer: ```suggestion
Getting to the end of the book, some publishers started to reach out to us. They wanted to publish Stratospheric from their publishing house and bring it to various marketplaces.  The additional outreach may be significant, but the royalties are fairly small compared to Leanpub, especially when split into three equal parts.
```

> To stay independent, we chose against a publisher and published the eBook ourselves on Amazon to have at least the big search traffic from this major e-commerce platform. We know that Leanpub is a niche site, and not all developers are aware of this great source of technical books. However, given our existing audience from our blog, newsletter, Twitter, etc., we make our potential readers aware via those channels.

Reviewer: ```suggestion
To stay independent, we opted against a publisher and published the eBook ourselves on Amazon to at least have  the enormous search traffic coming from this major e-commerce platform. We know that Leanpub is a niche site, and not all developers are aware of this great source of technical books. However, given our existing audience from our blog, newsletter, Twitter, etc. we can easily make our potential readers aware via those channels.
```

At the time of writing this article, we have 800 readers, which is a great success 🥳

## Recommendations, Tips & Tricks, and Lessons Learned 🎩

What follows is an unordered list of hints and tips about this journey. If you plan to start a similar project, make sure to learn from our _mistakes_ and recommendations.

### Write Together, Not Alone

Writing a book together has a lot of benefits. It motivates you to keep going when you would otherwise have quit when working on your own.

> We committed to delivering something every two weeks and had some high-level ideas regarding when the next release would be published. This goal-based work allowed us not to drift off and work towards the next milestone.

Reviewer: ```suggestion
We committed to delivering something every two weeks and had some high-level ideas regarding when the next release would be published. This goal-based approach allowed us not to veer off, but work towards the next milestone.
```

> Especially if it's your first book, writing it alone might feel like a big pile of work. Together, you get there step-by-step and can use the group's momentum.

Reviewer: ```suggestion
Especially if you're writing your first book, going it alone might feel like an insurmountable pile of work. Together, you get there step-by-step and can use the group's momentum.
```

Furthermore, when co-authoring a book, you'll also improve your writing skills by getting constant feedback from other co-authors.

> We all had some background in writing technical content, which helped us a lot. Starting such a project with three people who had never written any technical blog before might be burdensome.

Reviewer: ```suggestion
We all had some background in writing technical content, which helped us a lot. Starting such a project with three people who have never written any technical blog before might be burdensome.
```

### Use an Iterative Approach

With an iterative approach, you don't have to lock yourself for a year in a wooden cabin in Norway and come back with a finished book.

Start small. Deliver small increments, do baby steps, and get feedback early.

We had a rough plan of what we wanted to cover but also adjusted a lot along the way.

Leanpub is the perfect platform for this. Especially if you are a developer, you'll really enjoy writing your manuscript in Markdown. No annoying formatting issues inside Word when trying to center images or tables.

> Leanpub does the layout and file generation jobs for you. All you have to do is ship the manuscript.

Reviewer: ```suggestion
Leanpub does the layouting and file generation jobs for you. All you have to do is ship the manuscript.
```

> Pick a publisher that supports early access programs if you don't want to go towards the self-publishing route.

Reviewer: ```suggestion
Pick a publisher that supports early-access programs if you don't want to go down the self-publishing route.
```

You might be super convinced about your idea and think it's the most genius book ever. However, this doesn't necessarily reflect reality. Get some early feedback mechanisms in place that help you quantify the interest. A mailing list or a dedicated Twitter account can be such a tool.

### Use Simple Tools

Keep it simple. Use free plans and share the login with your co-authors if necessary.

You can scale the tools and costs with your eBook's success.

> We purchased a Bootstrap-based HTML theme for our landing page for some bucks and adjusted it to our needs. During the project's inception phase, and while the eBook was not at 100%, we primarily used the landing page to attract potential readers. Our landing page had (and still has) a newsletter subscription form to keep potential readers around.

Reviewer: ```suggestion
We purchased a Bootstrap-based HTML theme for our landing page for some bucks and adjusted it to our needs. During the project's inception phase, and while the eBook still wasn't at 100% completion, we primarily used the landing page to attract potential readers. Our landing page had (and still has) a newsletter subscription form to keep potential readers around.
```

> We started with the free plan of ConvertKit and recently migrated to MailLite. The MailerLite pricing and plan structure was better for our needs.

Reviewer: ```suggestion
We started with the free plan of ConvertKit and recently migrated to MailerLite. The MailerLite pricing and plan structure was better for our needs.
```

Some other tools/services that we used:
- Grammarly for fundamental proofreading and typos

> - Writing the eBook in their editor of choice. I personally used IntelliJ IDEA most of the time.  I tweaked the editor settings to wrap the lines so that I can write more naturally without having to scroll endlessly horizontally or buy a curved monitor.

Reviewer: ```suggestion
- Writing the eBook in our editor of choice. I personally used IntelliJ IDEA most of the time. I tweaked the editor settings to wrap the lines so that I can write more naturally without having to scroll endlessly horizontally or buy a curved monitor.
```

- Fiverr for designing creatives, the logo, and the eBook cover
- GitHub for collaboration and review
- Google Sheet to keep track of our costs
- Slack for internal and external (with the community) communication

### Find a Schedule to Keep Momentum

What helped us to keep the writing flow going was a fixed schedule. We met every two weeks for one hour and did a status update before planning the next steps.

What worked for us was:
- A Trello board with four columns: backlog, in progress, to review, done
- Meetings via Zoom
- Document writing standards within the repository
- Use GitHub Actions for basic automation

What personally worked for me was taking the first hour every morning to work on this project.

### Automate What Can Be Automated

> As we all had a software developer background, we tried to automate everything that we could. We could have also automated the eBook publishing using Leanpub's API.

Reviewer: ```suggestion
As we all had a software developer background and mindset, we tried to automate everything that we could. We could even have automated the eBook publishing process using Leanpub's API.
```

> As we all had a software developer background, we tried to automate everything that we could. We could have also automated the eBook publishing using Leanpub's API.

Reviewer: https://thethreevirtues.com/

GitHub Actions makes such automations almost effortless.

Adding links for further resources is both a curse and a blessing. It allows the interested reader to dive deeper into a topic. However, the links disappear over time and may no longer be accessible as the content might have moved, or the site may have been shut down.

We automated the hyperlink verification with a small GitHub Action and an npm package that checks all links (also internal manuscript links) on each commit:

```
name: Check for broken links
on: [push, pull_request]
jobs:
  broken-link-check:
    runs-on: ubuntu-latest
    name: Check for broken links in all Markdown files
    steps:
      - name: Check out source code
        uses: actions/checkout@v3

      - name: Set up Node 16
        uses: actions/setup-node@v3
        with:
          node-version: 16
          cache: 'npm'

      - name: Install NPM packages
        run: npm install

      - name: Check for broken links
        run: npm run check-links
```

### Reach Out for Advertisement

Start talking about your eBook on podcasts, conferences, and tweet about it as soon as possible.

> We reached out to an AWS developer advocate during the inception phase to find possible ways of collaboration.

Reviewer: ```suggestion
We reached out to an AWS developer advocate during the inception phase to find possible ways of collaborating.
```

> AWS has sponsored us with some gift vouchers so that we could run the infrastructure (almost) for free. We could even sponsor a two-part article on the AWS Open Source blog to gain some traction for our project:

Reviewer: ```suggestion
AWS has sponsored us with some gift vouchers so that we could run the infrastructure (almost) for free. We even were able to contribute a two-part article on the AWS Open Source blog to gain some traction for our project:
```

- [Getting started with Spring Boot on AWS: Part 1](https://aws.amazon.com/blogs/opensource/getting-started-with-spring-boot-on-aws-part-1/)
- [Getting started with Spring Boot on AWS: Part 2](https://aws.amazon.com/blogs/opensource/getting-started-with-spring-boot-on-aws-part-2/)

> The [German AWS Podcast](https://german-aws-podcast.libsyn.com/33-von-null-auf-produktion-mit-spring-boot-und-aws) from Dennis Traub (AWS Developer Advocate) was also a great initiative to spread the word.

Reviewer: ```suggestion
The [German AWS Podcast](https://german-aws-podcast.libsyn.com/33-von-null-auf-produktion-mit-spring-boot-und-aws) by Dennis Traub (AWS Developer Advocate) was a fantastic way of spread the word, too.
```

> After finishing the book, AWS even sponsored us for $2500 on GitHub.

Reviewer: ```suggestion
After finishing the book, AWS even sponsored us to the tune of $2500 on GitHub.
```

Reviewer: While more idiomatic, the expression I suggested is quite informal (and perhaps mostly used in British English). So, depending on the style you're aiming for, this might not be appropriate. The original simple preposition would be better in that case.

### Get an Email Address for Every Author

That's an optional recommendation, but we had a good experience with having a dedicated email for each co-author and a basic catch-all email like `info@yourbookname.com`.

We've registered our domain within Amazon Route 53. We've built a simple self-made email forwarding with Amazon SES, S3, and AWS Lambda. If you're curious about the implementation, check out the [source code on GitHub](https://github.com/stratospheric-dev/stratospheric/tree/main/utility/email).

### Work With Checklists

Whenever there was something that we couldn't or didn't want to automate, we created a checklist for it.

> This way, we didn't forget a step for tasks that we only occasionally performed.

Reviewer: ```suggestion
This way, we didn't forget a step for tasks we only performed occasionally.
```

As an example, this is our checklist for publishing a new eBook version:

![Stratospheric Sample Checklist](/img/stratospheric-ebook/stratospheric-sample-check-list-example.png#center "Stratospheric Sample Checklist")

### Rotate the Responsibilities

To avoid a single point of failure within the co-author team and mix things up, we rotated the various responsibilities.

Everyone should release a new eBook version every now and then.

> We also have an "author of the week" that keeps an eye on our feedback channels and is the first responder and coordinator for GitHub issues, Slack notifications, and emails.

Reviewer: ```suggestion
We also have an "author of the week" keeping an eye on our feedback channels. That person also is the first responder and coordinator for GitHub issues, Slack notifications, and emails.
```

### Centralize the Feedback Channels

Right after the first readers got their copy, we started to get an increasing amount of messages via various channels:

- private DMs on Twitter
- Emails to our eBook email address and/or private email address
- Slack messages
- Feedback via Leanpub
- GitHub issues
- etc.

> The various channels made it hard for us to keep track of all inquiries. Coordinating the response among the co-author is difficult if someone gets a private Twitter DM.

Reviewer: ```suggestion
The various channels made it hard for us to keep track of all inquiries. Coordinating the response amongst the co-authors is difficult if someone gets a private Twitter DM.
```

    validations:
      required: true
  - type: input
    id: version
    attributes:
      label: Version
      description: "Which eBook version are you referring to? You can find the version as part of the changelog (one of the last pages). Pick 'latest' if you can't find it."
    validations:
      required: true
  - type: textarea
    id: suggested-change
    attributes:
      label: Suggested Changes
      description: What would you suggest to change or what do you expect as a reader?
```

The rendered version of this template looks like the following:

![Centralized Feedback via GitHub Issues](/img/stratospheric-ebook/stratospheric-ebook-create-github-issue.png#center "Centralized Feedback via GitHub Issues")

Whenever a reader now asks a question via any other channel, we redirect them to [GitHub](https://github.com/stratospheric-dev/stratospheric/issues) and ask them to submit their message there.

### Define Rules for Typesetting and Structural Elements

This would have helped us save some time for the final polishing and is one of our main lessons learned.

> Three different authors have three different opinions on when to use bold or italics. Sometimes there are even four opinions if you forget which rules you applied yesterday.

Reviewer: ```suggestion
Three different authors have three different opinions on when to use bold or italic script. Sometimes there are even four opinions if you forget which rules you applied yesterday.
```

Examples of this include:

- When to use *bold*?
- When to use _italics_?

> - When to use an infobox?

Reviewer: ```suggestion
- When to use an info box?
```

- How to structure the headers?

> - How and which screenshot to take?

Reviewer: ```suggestion
- How to take a screenshot and which screenshot to take?
```

- What is an acceptable number of links on a book page?
- etc.

We did not cover and align those rules in detail during our early manuscript reviews. During the final proofreading session, we encountered the issue of non-consistent sentence highlighting.

As a result, we now have some basic standards on when and how to highlight sections or single words.

> If you're starting from scratch, take some time with your authors to define clear definitions and create a one-pager for your typesetting rules.

Reviewer: ```suggestion
If you're starting from scratch, take some time with your co-authors to set clear definitions and create a one-pager for your typesetting rules.
```

### Have Fun

> That's the most important takeaway. Find people that you enjoy working with and that you can learn from.

Reviewer: ```suggestion
That's the most important takeaway. Find people you enjoy working with and from whom you can learn.
```

It's a side project, enjoy the experience, and learn as much as you can. It's your spare time. If it feels like a job, don't do it.

> Don't expect to become rich by writing a book but instead, use this opportunity to establish a brand in your niche or use it as a form of credibility. I'm pretty sure that co-authoring an eBook weighs more than any basic certification for a specific topic.

Reviewer: ```suggestion
Don't expect to become rich by writing a book but instead, use this opportunity to establish a brand in your niche or use it as a way of establishing credibility and authority. I'm fairly certain co-authoring an eBook is more relevant than any basic certification for a specific topic.
```

Find out more about Tom on [reflectoring.io](https://reflectoring.io) and his [Twitter profile](https://twitter.com/TomHombergs).

### Björn Wilmsmann

Björn Wilmsmann is an independent IT consultant who helps companies transform their business into a digital business. He designs and develops business solutions and enterprise applications for his clients. Björn provides hands-on training in technologies such as Angular and Spring Boot.

Find out more about Björn on [bjoernkw.com](https://bjoernkw.com) and his [Twitter profile](https://twitter.com/bwilmsmann).

### Philip Riecks (That's Me)

Under the slogan _Testing Java Applications Made Simple_ Philip provides recipes and tips & tricks to accelerate your testing success and make testing joyful (or at least less painful). Apart from blogging, he's a course instructor for various Java-related online courses and is active on YouTube.

Find out more about Philip on [rieckpil.de](https://rieckpil.de) and his [Twitter profile](https://twitter.com/rieckpil).

If you have further questions or want some more details about a specific topic, let me know in the comments 👇🏻

> PS: You can get a reduced copy of the Stratospheric eBook [here](https://stratospheric.dev/).

Reviewer: ```suggestion
PS: You can get a copy of the Stratospheric eBook at a discount [here](https://stratospheric.dev/).
```
